### PR TITLE
Allow forwarding refs on SelectMenu

### PR DESCRIPTION
**Pull request by emplums**
This PR updates `SelectMenu` to forward refs. I used the `useProvidedRefOrCreate` approach suggested by @smockle ✨ 

Closes: https://github.com/github/design-systems/issues/857

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -413,6 +414,7 @@ declare module '@primer/components' {
     block?: boolean
     icon?: ReactComponentLike
     variant?: 'small' | 'large'
+    ref?: HTMLSpanElement
```
Comment (emplums) [inline index.d.ts:417]: @smockle does this type definition look right??

Comment (smockle) [reply index.d.ts]: I think we discussed this as part of https://github.com/github/memex/pull/1171, and I note the latest commit in this PR uses a different type, so I’m going to mark this conversation as “resolved”. If there is more to discuss, though, please feel free to re-open! 😊

**Comment by emplums** [comment]
I've been using this in ✨ another project ✨  all week and it seems to be working well!

File: src/SelectMenu/SelectMenuModal.js [hunk 2]
```diff
@@ -29,14 +29,14 @@ const modalStyles = css`
...
   @media (min-width: ${get('breakpoints.0')}) {
-    width: '300px';
+    width: 300px;
     height: auto;
     max-height: 350px;
     margin: ${get('space.1')} 0 ${get('space.3')} 0;
     font-size: ${get('fontSizes.0')};
     border: ${get('borderWidths.1')} solid ${get('colors.border.grayDark')};
     border-radius: ${get('radii.2')};
-    box-shadow: 0 1px 5px ${get('colors.blackfade15')} !default;
```
Comment (emplums) [inline src/SelectMenu/SelectMenuModal.js:39]: This is invalid CSS obvs, I think it was supposed to be `!important` but it doesn't look like that was needed either so I removed it

**Review by BinaryMuse — APPROVED** [review]
Looks good! My only question is about versioning. From https://github.com/github/github/pull/157403:
> ## Note for component library maintainers
> 
> **When you start using forwardRef in a component library, you should treat it as a breaking change and release a new major version of your library.** This is because your library likely has an observably different behavior (such as what refs get assigned to, and what types are exported), and this can break apps and other libraries that depend on the old behavior.

File: src/SelectMenu/SelectMenu.js [hunk 3]
```diff
@@ -51,8 +51,9 @@ const StyledSelectMenu = styled.details`
 `
 
 // 'as' is spread out because we don't want users to be able to change the tag.
-const SelectMenu = ({children, initialTab, as, ...rest}) => {
-  const ref = useRef(null)
+const SelectMenu = React.forwardRef(({children, initialTab, as, ...rest}, forwardedRef) => {
+  const backupRef = useRef()
```
Comment (BinaryMuse) [inline src/SelectMenu/SelectMenu.js:55]: Based on https://reactjs.org/docs/forwarding-refs.html#note-for-component-library-maintainers, could we give this function a name so it shows up as expected in devtools?

```suggestion
const SelectMenu = React.forwardRef(
  function SelectMenu({children, initialTab, as, ...rest}, forwardedRef) {
    const backupRef = useRef()
```

**Comment by emplums** [comment]
@BinaryMuse I believe because we are using a backup ref, it would be ok to not version this as a breaking change since omitting a ref here will not break any functionality!

File: src/SelectMenu/SelectMenu.js (hunk 3 above)
Comment (emplums) [reply src/SelectMenu/SelectMenu.js:55]: The name seems to show up fine in devTools! 
image

**Pull request merged**